cmd/util/ledger/migrations/cadence_value_validation.go

```diff
@@ -0,0 +1,496 @@
+package migrations
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onflow/atree"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/ledger"
+)
+
+var nopMemoryGauge = util.NopMemoryGauge{}
+
+// TODO: optimize memory by reusing payloads snapshot created for migration
+func validateCadenceValues(
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+	newPayloads []*ledger.Payload,
+) error {
+	// Create all the runtime components we need for comparing Cadence values.
+	oldRuntime, err := newReadonlyStorageRuntime(oldPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with old payloads: %w", err)
+	}
+
+	newRuntime, err := newReadonlyStorageRuntime(newPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with new payloads: %w", err)
+	}
+
+	// Iterate through all domains and compare cadence values.
+	for _, domain := range domains {
```

Comment:
Can we let the caller pass the domains as an argument?

Reply:
Good idea but in this case, it's not needed because we only deal with a hard coded set of domains defined in migration PR 4633.